scipy/interpolate/regrid_python.py

```diff
@@ -0,0 +1,1104 @@
+    s : float
+        Smoothing target (> 0 for smoothing; if s == 0, do pure interpolation).
+    nmin, nmax : int, optional
+        Lower/upper bounds on knot count. If t is None, these are derived as:
+          nmin = 2*(k+1), nmax = m + k + 1 (m = x.size).
+    nest : int, optional
+        Storage cap for knots. If t is None and nest is None, it defaults to
+        max(m + k + 1, 2*k + 3). Must satisfy nest >= 2*(k+1).
+    t : 1-D ndarray, optional
+        Current knot vector. If None, the function initializes t as
+        [xb]*(k+1) + [xe]*(k+1).
+    fp, fpold : float, optional
+        Current and previous residual sums of squares (needed after initialization).
+    residuals : 1-D ndarray, optional
+        Most recent residual signal used by add_knot to place the next knot.
+    nplus : int, optional
+        Previous iteration's proposed number of knots to add (used in nplus update).
+
+    Returns
+    -------
+    On first call (t is None):
+        t, nest, nmin, nmax
+            Initialized knot vector and derived limits.
+    On acceptance (|fp - s| < acc or fp < s):
+        None
+            Caller should stop; current t is acceptable.
+    On growth step:
+        t, nplus
+            Updated knot vector after inserting up to nplus knots;
+            also the nplus chosen.
+    On early stop due to n >= nmax or n >= nest:
+        t, nplus   (or just t depending on branch)
+            Final knot vector respecting the cap.
+
+    Notes
+    -----
+    - Use this as a building block in an outer loop:
+        1) Call with t=None to initialize (and receive nest, nmin, nmax).
+        2) Fit/evaluate to compute fp and residuals for current t.
+        3) Call again with updated fp, fpold, residuals to either accept, or
+           get back an updated t (and nplus) to continue.
+    - For s == 0, this routine returns a not-a-knot vector immediately and
+      expects the caller to perform interpolation.
+    - This helper mirrors FITPACK's growth heuristics but intentionally leaves
+      residual computation and periodic handling to the caller/higher level.
+    """
+
+    if s == 0:
+        if nest is not None:
+            raise ValueError("s == 0 is interpolation only")
+        # For special-case k=1 (e.g., Lyche and Morken, Eq.(2.16)),
+        # _not_a_knot produces desired knot vector
+        t = _not_a_knot(x, k)
+        return np.asarray(t), None, None, None
+
+    acc = s * TOL
+    m = x.size    # the number of data points
+
+    if t is None:
+        if nest is None:
+            # the max number of knots. This is set in _fitpack_impl.py line 274
+            # and fitpack.pyf line 198
+            # Ref: https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/_fitpack_impl.py#L260-L263
+            nest = max(m + k + 1, 2*k + 3)
+        else:
+            if nest < 2*(k + 1):
+                raise ValueError(f"`nest` too small: {nest = } < 2*(k+1) = {2*(k+1)}.")
```

Comment:
This should be handled once, before entering the main loop. In fact, I'd handle the whole `t is None` clause early on, before the main loop.